lib/common/socketpool.c

```diff
@@ -441,6 +386,89 @@ static void start_connect(h2o_socketpool_connect_request_t *req, struct sockaddr
     req->sock->on_close.data = close_data;
 }
 
+static void try_connect(h2o_socketpool_connect_request_t *req)
+{
+    h2o_socketpool_target_t *target;
+    h2o_socketpool_t *pool = req->pool;
+    struct pool_entry_t *entry = NULL;
+    struct on_close_data_t *close_data;
+    h2o_linklist_t *sockets = NULL;
+    
+    req->remaining_try_count--;
+
+    if (req->lb.tried != NULL) {
+        if (req->pool->targets.size > 1) {
+            req->selected_target = req->pool->balancer->callbacks->selector(req->pool->balancer, &req->pool->targets,
+                                                                      req->lb.tried, req->lb.req_info);
+            assert(!req->lb.tried[req->selected_target]);
+            req->lb.tried[req->selected_target] = 1;
+            __sync_add_and_fetch(&pool->targets.entries[req->selected_target]->_shared.leased_count, 1);
+        } else {
+            req->selected_target = 0;
+        }
+    }
+    target = req->pool->targets.entries[req->selected_target];
+    sockets = &pool->targets.entries[req->selected_target]->_shared.sockets;
+
+    pthread_mutex_lock(&pool->_shared.mutex);
+    /* try to fetch an entry and return it */
```

Comment:
To me it seems that the semantics has changed here.

Previously, we were first looking for a pool socket and reusing it. Whereas now the code looks for a target then selects a pool socket.

Am I understanding correctly? If so, could you please elaborate why such change is needed?

Reply:
Sure. Such change is needed for ensuring that balancing is applied on each request. If not, considering upstreams support presistent connections so that sockets used would be returned to socketpool. We then cannot control which target would be used when new request comes. If there're many sockets pooled with each upstream and suddenly serveral upstreams go down then back online, load could be unbalanced, i.e. those recovered ones would get fewer requests than others or even none if sockets returned are enough for new requests, especially when balancer is naive round robin.

Reply:
Thank you for the explanation.

> If there're many sockets pooled with each upstream and suddenly serveral upstreams go down then back online, load could be unbalanced, i.e. those recovered ones would get fewer requests than others or even none if sockets returned are enough for new requests, especially when balancer is naive round robin.

I agree with the observation. OTOH, I wonder if per-request load balancing is the correct answer here.

When a new node is added as a backend, it is often a good idea to gradually increase the percentage of the requests handled by the node (i.e. warmup). Current implementation guarantees that as a side effect of selecting the nodes only when new connections are established. How fast rebalance completes depends on how many requests each backend handles before closing the connection. Switching to per-request balancing removes that kind of operational benefit.

It should also be noted that connection establishment is not zero-cost; it could take as long as 3-RTT (i.e. TCP SYN + TLS handshake). TLS handshake is a heavyweight operation in terms of CPU cycles. So when connections to the backends are established via WAN, it is beneficial to prefer reusing existing connections than strictly balancing the load.

Considering the aspects, I would appreciate it if you could go with per-connection balancing in this PR, and separately start a discussion about if (and how) we should support per-request balancing (if that is required).